### PR TITLE
New Release Process

### LINKED ISSUE #290: Release Pulsar Helm Chart on dist.apache.org
## Motivation

Currently, we host Pulsar Helm Chart's metadata (`index.yaml`) on https://pulsar.apache.org/charts/index.yaml. Users encountered an issue today (https://github.com/apache/pulsar/issues/17799) due to we removed the file during "clean up" content before building the official website fully.

Depending on another repo is brittle. It's hard to ensure further development happening in the pulsar-site repo is aware of this trick while this is critical to the pulsar-helm-chart usage.

## Proposal

I propose we adopt releasing Pulsar Helm Chart in GitHub Pages flavor to make this chart self-contained.

* Helm official document: https://helm.sh/docs/howto/chart_releaser_action/
* Chart release action: https://github.com/helm/chart-releaser-action
* A live example from OpenSearch:
    * https://github.com/opensearch-project/helm-charts/blob/main/.github/workflows/release.yaml
    * https://github.com/opensearch-project/helm-charts/blob/gh-pages/index.yaml
* Prototype on fork: https://github.com/tisonkun/pulsar-helm-chart

## Implementation

The implementation steps are:

1. Create `gh-pages` branch on pulsar-helm-chart repo, add current `index.yaml` file to the branch.
2. Ask Apache INFRA team to config the branch as the GitHub Pages branch.
3. Add the workflow using chart-releaser-action (bundled in this repo, it's lightweight).
4. Redirect https://pulsar.apache.org/charts to the new place.

## Risks

~~The last step _may_ be optional because IIUC helm chart's index is global. Thus once we publish the first new index with chart-releaser-action, all downloads to the chart will find the index from the new place. Although, I'm not an expert on Helm Chart so here requires some verification.~~ No. `https://pulsar.apache.org/charts` is actually documented at https://pulsar.apache.org/docs/helm-overview/#upgrading. So we should make the redirection.

Another risk is that the reliability of pulsar.apache.org is promised by the ASF INFRA team, while GitHub Pages are handled by the GitHub team. I'd say GitHub is less stable than ASF sites.

## Alternative

Perform an Apache release and upload the index to https://dist.apache.org/repos/dist/release/pulsar. This approach requires modifying homemade automation (dist.a.o uses SVN), and possibly brings new bugs.

### PR DESCRIPTION
**Pull request by michaeljmarshall**
Relates to #290

### Motivation

Make the Apache Pulsar Helm Chart release follow ASF rules for voting, and make the helm binary available via dist.apache.org. By following the information in https://issues.apache.org/jira/browse/LEGAL-573 and in the Apache Airflow project https://github.com/apache/airflow/blob/main/dev/README_RELEASE_HELM_CHART.md, I built this new release process. It will likely need some iterative improvement.

### Modifications

* Add a release process that is based on the Apache Airflow release process

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

### CODE REVIEW

File: RELEASE.md [hunk 1]
```diff
@@ -0,0 +1,576 @@
...
+ software distributed under the License is distributed on an
+ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ KIND, either express or implied.  See the License for the
+ specific language governing permissions and limitations
+ under the License.
+-->
+**Table of contents**
+
+- [Prepare the Apache Pulsar Helm Chart Release Candidate](#prepare-the-apache-pulsar-helm-chart-release-candidate)
+  - [Pre-requisites](#pre-requisites)
```
Comment (maxsxu) [inline RELEASE.md:22]:
```suggestion
  - [Prerequisites](#prerequisites)
```

File: RELEASE.md [hunk 2]
```diff
@@ -0,0 +1,576 @@
...
+  - [Close the milestone](#close-the-milestone)
+  - [Announce the release on the community slack](#announce-the-release-on-the-community-slack)
+  - [Bump chart version in Chart.yaml](#bump-chart-version-in-chartyaml)
+  - [Remove old releases](#remove-old-releases)
+
+This document details the steps for releasing the Apache Pulsar Helm Chart.
+
+# Prepare the Apache Pulsar Helm Chart Release Candidate
+
+## Pre-requisites
```
Comment (maxsxu) [inline RELEASE.md:47]:
```suggestion
## Prerequisites
```

**Review by maxsxu — APPROVED** [review]

File: RELEASE.md [hunk 3]
```diff
@@ -0,0 +1,576 @@
...
+- [Publish the final release](#publish-the-final-release)
+  - [Summarize the voting for the release](#summarize-the-voting-for-the-release)
+  - [Publish release to SVN](#publish-release-to-svn)
+  - [Publish release tag](#publish-release-tag)
+  - [Notify developers of release](#notify-developers-of-release)
+  - [Create release on GitHub](#create-release-on-github)
+  - [Close the milestone](#close-the-milestone)
+  - [Announce the release on the community slack](#announce-the-release-on-the-community-slack)
+  - [Bump chart version in Chart.yaml](#bump-chart-version-in-chartyaml)
+  - [Remove old releases](#remove-old-releases)
```
Comment (tisonkun) [inline RELEASE.md:41]: As mentioned in https://github.com/apache/pulsar/pull/17470#discussion_r963277120, we don't have to manually maintain a TOC section.

Comment (tisonkun) [reply RELEASE.md:41]:
```suggestion
```

File: RELEASE.md [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+# Prepare the Apache Pulsar Helm Chart Release Candidate
+
+## Prerequisites
+
+- Helm version >= 3.0.2
+- Helm gpg plugin (one option: https://github.com/technosophos/helm-gpg)
+
+## Build Release Notes
+
+Before creating the RC, you need to build and commit the release notes for the release.
```
Comment (tisonkun) [inline RELEASE.md:54]: How to build and where to commit?

File: RELEASE.md [hunk 5]
```diff
@@ -0,0 +1,576 @@
...
+
+## Build RC artifacts
+
+The Release Candidate artifacts we vote upon should be the exact ones we vote against,
+without any modification than renaming – i.e. the contents of the files must be
+the same between voted release candidate and final release.
+Because of this the version in the built artifacts that will become the
+official Apache releases must not include the rcN suffix.
```
Comment (tisonkun) [inline RELEASE.md:62]: I suggest we don't do manual line wrap. But this is not a hard requirement :)

File: RELEASE.md [hunk 6]
```diff
@@ -0,0 +1,576 @@
...
+    export VERSION=1.0.1-candidate-1
+    export VERSION_WITHOUT_RC=${VERSION%-candidate-*}
+
+    # Set PULSAR_REPO_ROOT to the path of your git repo
+    export PULSAR_REPO_ROOT=$(pwd)
+
+    # Example after cloning
+    git clone https://github.com/apache/pulsar-helm-chart.git pulsar
+    cd pulsar-helm-chart
+    export PULSAR_REPO_ROOT=$(pwd)
```
Comment (tisonkun) [inline RELEASE.md:77]:
```suggestion
    # Clone and set PULSAR_REPO_ROOT
    git clone https://github.com/apache/pulsar-helm-chart.git pulsar
    cd pulsar-helm-chart
    export PULSAR_REPO_ROOT=$(pwd)
```

I think for those who already have a clone, it should be intuitive to adjust.

File: RELEASE.md [hunk 7]
```diff
@@ -0,0 +1,576 @@
...
+- We currently release Helm Chart from `master` branch:
+
+    ```shell
+    git checkout master
+    ```
+
+- Clean the checkout: the sdist step below will
+
+    ```shell
+    git clean -fxd
```
Comment (tisonkun) [inline RELEASE.md:89]:
```suggestion
    git clean -fdX .
```

We should use the ignorance rules by `.gitignore` file?

File: RELEASE.md [hunk 8]
```diff
@@ -0,0 +1,576 @@
...
+  mv ${PULSAR_REPO_ROOT}/pulsar-chart-${VERSION_WITHOUT_RC}-source.tar.gz* ${VERSION}/
+  cd ${VERSION}
+
+  ###### Generate index.yaml file - Start
+  # Download the latest index.yaml on Pulsar Website
+  curl https://pulsar.apache.org/charts/index.yaml --output index.yaml
+
+  # Replace the URLs from "https://downloads.apache.org" to "https://archive.apache.org"
+  # as the downloads.apache.org only contains latest releases.
+  sed -i 's|https://downloads.apache.org/pulsar/helm-chart/|https://archive.apache.org/dist/pulsar/helm-chart/|' index.yaml
```
Comment (tisonkun) [inline RELEASE.md:181]: Shall we upload all previous releases to `archive.apache.org`?

File: RELEASE.md [hunk 9]
```diff
@@ -0,0 +1,576 @@
...
+<TODO COPY LINK TO THE ISSUE CREATED>
+
+Thanks,
+<your name>
+EOF
+```
+
+Note, you need to update the `helm gpg verify` output and verify the end of the voting period in the body.
+
+# Verify the release candidate by PMCs
```
Comment (tisonkun) [inline RELEASE.md:289]:
```suggestion
# Verify the release candidate by the PMC
```

PMC refers to the Project Management Committee.

File: RELEASE.md [hunk 10]
```diff
@@ -0,0 +1,576 @@
...
+
+The PMCs should verify the releases in order to make sure the release is following the
```
Comment (tisonkun) [inline RELEASE.md:291]:
```suggestion
The PMC should verify the releases in order to make sure the release is following the
```

File: RELEASE.md [hunk 11]
```diff
@@ -0,0 +1,576 @@
...
+[Apache Legal Release Policy](http://www.apache.org/legal/release-policy.html).
+
+At least 3 (+1) votes should be recorded in accordance to
```
Comment (tisonkun) [inline RELEASE.md:294]:
```suggestion
At least 3 (+1) votes from PMC members should be recorded in accordance to
```

File: RELEASE.md [hunk 12]
```diff
@@ -0,0 +1,576 @@
...
+
+## Summarize the voting for the release
+
+Once the vote has been passed, you will need to send a result vote to dev@pulsar.apache.org:
+
+Subject:
+
+```
+[RESULT][VOTE] Release Apache Pulsar Helm Chart 1.0.1 based on 1.0.1-candidate-1
+```
```
Comment (tisonkun) [inline RELEASE.md:428]: Shall we also use shell and replace placeholder `1.0.1` with variables here?

File: RELEASE.md [hunk 13]
```diff
@@ -0,0 +1,576 @@
...
+
+# Create new folder for the release
+cd pulsar-dist-release/helm-chart
+export PULSAR_SVN_RELEASE_HELM=$(pwd)
+svn mkdir ${VERSION}
+cd ${VERSION}
+
+# Move the artifacts to svn folder & commit (don't copy or copy & remove - index.yaml)
+for f in ../../../pulsar-dist-dev/helm-chart/$RC/*; do svn cp $f ${$(basename $f)/}; done
+svn rm index.yaml
```
Comment (tisonkun) [inline RELEASE.md:476]: In this release manual, it doesn't talk about updating the `index.yaml` file on https://pulsar.apache.org/charts/index.yaml, and here we remove the `index.yaml` file on dist.a.o

Then where do we host the `index.yaml` file?

I suggest we host it on the dist.a.o and redirect https://pulsar.apache.org/charts/index.yaml to there.

**Review by tisonkun — CHANGES_REQUESTED** [review]
Thanks for preparing the release process @michaeljmarshall!

Comments inline.

File: RELEASE.md (hunk 9 above)
Comment (michaeljmarshall) [reply RELEASE.md:289]: Good catch.

File: RELEASE.md (hunk 4 above)
Comment (michaeljmarshall) [reply RELEASE.md:31]: I haven't really defined this yet. The helm chart doesn't move that quickly, so at this point, I was thinking it would be manual. Apache Airflow has a script to generate this information: https://github.com/apache/airflow/blob/main/dev/chart/build_changelog_annotations.py.

File: RELEASE.md (hunk 5 above)
Comment (michaeljmarshall) [reply RELEASE.md:39]: I only had manual line wraps because I copied the Airflow process and wanted to provide reviewers a meaningful diff. I don't feel strongly either way.

File: RELEASE.md (hunk 8 above)
Comment (michaeljmarshall) [reply RELEASE.md:155]: The `index.yaml` can point to multiple locations, so I don't think we need to update the ones hosted on GitHub. Primarily, the ones on GitHub were not voted on by the Apache Pulsar PMC, so I don't think we should move them to `archive.apache.org`. I don't have a strong opinion though.

Comment (tisonkun) [reply RELEASE.md:155]: Yes. I notice that this command only affects the Apache releases. Than it looks good to me :)

File: RELEASE.md (hunk 12 above)
Comment (michaeljmarshall) [reply RELEASE.md:402]: Yes, we could definitely do that. As I revised the Apache Airflow documentation, I felt like we could make a lot of this a script that outputs different content into files for the release manager to then upload or share in the appropriate way.

File: RELEASE.md (hunk 13 above)
Comment (michaeljmarshall) [reply RELEASE.md:450]: Great point, I will take a look at how we update the index.yaml tomorrow. I support with hosting it on dist.a.o.

**Comment by michaeljmarshall** [comment]
Thanks for your thorough review @tisonkun. I will work on improving the doc tomorrow (my time).

File: RELEASE.md (hunk 12 above)
Comment (tisonkun) [reply RELEASE.md:402]: Than let's do it in following steps to write scripts :)

If this patch merged, we can file an issue for it.

File: RELEASE.md (hunk 13 above)
Comment (tisonkun) [reply RELEASE.md:450]: The official [helm/chart-releaser](https://github.com/helm/chart-releaser) may help.

Basically, we port the automation proposed in #292 with manual steps.

Comment (michaeljmarshall) [reply RELEASE.md:450]: Makes sense. I am going to merge some final PRs today and then start testing out the workflow to see how it would work. It will likely require some tweaking.

**Comment by michaeljmarshall** [comment]
I am going to merge this new process now, and we'll tweak it as needed.

File: RELEASE.md (hunk 12 above)
Comment (michaeljmarshall) [reply RELEASE.md]: #325

File: RELEASE.md (hunk 13 above)
Comment (tisonkun) [reply RELEASE.md:441]: Could you summarize a bit what actions we perform against the `index.yaml` file?

I can see that we remove here and add in the steps below, but it's not quite straightforward to review - a flow about where and what we do with the `index.yaml` file helps.

**Review by tisonkun — APPROVED** [review]
SGTM. One comments inline.

File: RELEASE.md (hunk 13 above)
Comment (michaeljmarshall) [reply RELEASE.md:441]: That is fair criticism. I'll document the flow when I get to that step next week.
